src/Nethermind/Nethermind.JsonRpc/Data/TransactionForRpc.cs

```diff
@@ -36,7 +36,14 @@ public TransactionForRpc(Keccak? blockHash, long? blockNumber, int? txIndex, Tra
             MaxFeePerGas = transaction.MaxFeePerGas;
             MaxPriorityFeePerGas = transaction.MaxPriorityFeePerGas;
         }
-        ChainId = transaction.ChainId;
+        if (transaction.Type > TxType.Legacy)
+        {
+            ChainId = transaction.ChainId ?? BlockchainIds.Mainnet;
+        }
+        else
+        {
+            ChainId = transaction.ChainId;
+        }
```

Comment:
Only relevant change in this PR: if the Transaction is not Legacy, then `ChainId` should not be null, so we default to `Mainnet` if so.

Reply:
It should default to one from chainspec, not Mainnet

Reply:
Defaulting to the `ChainId` on `ChainSpec` would make this change extremely cumbersome since we have:

- 38 usages of `TransactionForRpc(Transaction transaction)`
- 23 usages of `public TransactionForRpc(Keccak? blockHash, long? blockNumber, int? txIndex, Transaction transaction, UInt256? baseFee = null)`.

Each usage might or not have access to it. Are we sure that we need to do it this way?

Reply:
yeah it sucks, maybe do a static one in `BlockchainIds` that would be set when chainspec is loaded?

Reply:
Done. Not the prettiest solution, but any other would require a very large amount of changes (TODO for a future refactor)